apisix/discovery/nacos.lua

```diff
@@ -0,0 +1,319 @@
+    log.info("connect_timeout:", connect_timeout, ", send_timeout:", send_timeout,
+             ", read_timeout:", read_timeout, ".")
+    httpc:set_timeouts(connect_timeout, send_timeout, read_timeout)
+    local res, err = httpc:request_uri(url, {
+        method = method,
+        headers = headers,
+        body = body,
+        ssl_verify = true,
+    })
+    if not res then
+        return nil, err
+    end
+
+    if not res.body or res.status ~= 200 then
+        return nil, "status = " .. res.status
+    end
+
+    local json_str = res.body
+    local data, err = core.json.decode(json_str)
+    if not data then
+        return nil, err
+    end
+    return data
+end
+
+
+local function get_url(request_uri, path)
+    return request(request_uri, path, nil, "GET", nil)
+end
+
+
+local function post_url(request_uri, path, body)
+    return request(request_uri, path, body, "POST", nil)
+end
+
+
+local function get_token_param(base_uri, username, password)
+    if username and password then
+        local data, err = post_url(base_uri, auth_path .. "?username=" .. username
+                                   .. "&password=" .. password, nil)
+        if err then
+            log.error("nacos login fail:" .. username .. " " .. password .. " desc:" .. err)
+            return nil, err
+        end
+        return "&accessToken=" .. data.accessToken
+    end
+    return ""
+end
+
+
+local function get_base_uri()
+    local host = local_conf.discovery.nacos.host
+    -- TODO Add health check to get healthy nodes.
+    local url = host[math_random(#host)]
+    local auth_idx = str_find(url, "@")
+    local username, password
+    if auth_idx then
+        local protocol_idx = str_find(url, "://")
+        local protocol = string_sub(url, 1, protocol_idx + 2)
+        local user_and_password = string_sub(url, protocol_idx + 3, auth_idx - 1)
+        local arr = ngx_re.split(user_and_password, ":")
+        if #arr == 2 then
+            username = arr[1]
+            password = arr[2]
+        end
+        local other = string_sub(url, auth_idx + 1)
+        url = protocol .. other
+    end
+
+    if local_conf.discovery.nacos.prefix then
+        url = url .. local_conf.discovery.nacos.prefix
+    end
+
+    if str_byte(url, #url) ~= str_byte("/") then
+        url = url .. "/"
+    end
+
+    return url, username, password
+end
+
+
+local function get_page_service(infos, base_uri, token_param, page_num)
+    local path = str_format(service_list_path, page_num, 100) .. token_param
+    local data, err = get_url(base_uri, path)
+    if err then
+        return data, err, path
+    end
+
+    for _, service_name in ipairs(data.doms) do
+        core.table.insert(infos, service_name)
+    end
+    return data, err, path
+end
+
+
+local function iter_and_add_service_info(infos, base_uri, token_param)
+    local data, err, path = get_page_service(infos, base_uri, token_param, 1)
+    if err then
+        log.error("get_url:" .. path .. " err:" .. err)
```

Comment:
bad code style

this is the right style: `log.error("get_url:", path, " err:", err)`